File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/core/clock.rb
# frozen_string_literal: true

require "monitor"

module IAIso
  module Core
    # Source of seconds-since-epoch as a Float. This is documentation;
    # any object responding to `now` works (Ruby duck typing).
    module Clock
      def now
        raise NotImplementedError
      end
    end

    # Real time via Time.now.
    class WallClock
      INSTANCE = new.freeze

      def now = Time.now.to_f

      def self.instance = INSTANCE
    end

    # Pre-recorded clock values; deterministic for tests.
    class ScriptedClock
      include MonitorMixin

      def initialize(sequence)
        super()
        @sequence = sequence.map(&:to_f)
        @idx = 0
      end

      def now
        synchronize do
          return 0.0 if @sequence.empty?
          v = @idx < @sequence.size ? @sequence[@idx] : @sequence.last
          @idx += 1 if @idx < @sequence.size
          v
        end
      end

      def reset
        synchronize { @idx = 0 }
      end
    end

    # Block-backed clock.
    class ClosureClock
      def initialize(&block)
        @block = block
      end

      def now = @block.call.to_f
    end
  end
end

File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoCore/Clock.swift
import Foundation

/// Source of double-precision seconds since the epoch.
public protocol Clock: Sendable {
    func now() -> Double
}

/// Real time via `Date().timeIntervalSince1970`.
public final class WallClock: Clock {
    public static let shared = WallClock()
    private init() {}
    public func now() -> Double {
        return Date().timeIntervalSince1970
    }
}

/// Pre-recorded clock values; deterministic for tests.
public final class ScriptedClock: Clock, @unchecked Sendable {
    private let lock = NSLock()
    private var idx = 0
    private let sequence: [Double]

    public init(_ sequence: [Double]) {
        self.sequence = sequence
    }

    public func now() -> Double {
        lock.lock()
        defer { lock.unlock() }
        if sequence.isEmpty { return 0 }
        let i = idx
        idx += 1
        return i < sequence.count ? sequence[i] : sequence[sequence.count - 1]
    }

    public func reset() {
        lock.lock()
        defer { lock.unlock() }
        idx = 0
    }
}

/// Closure-backed clock.
public final class ClosureClock: Clock, @unchecked Sendable {
    private let fn: @Sendable () -> Double
    public init(_ fn: @escaping @Sendable () -> Double) { self.fn = fn }
    public func now() -> Double { fn() }
}
